### PR TITLE
Flow service: Add Blob support

### PR DESCRIPTION
**Pull request by kjcho-msft**

### CODE REVIEW

File: Services/DataX.Config/DataX.Config.Input.EventHub/Processor/CreateEventHubConsumerGroup.cs [hunk 1]
```diff
@@ -77,6 +77,31 @@ public override async Task<FlowGuiConfig> HandleSensitiveData(FlowGuiConfig guiC
                 guiConfig.Input.Properties.InputResourceGroup = secretId;
             }
 
+
+            if (guiConfig?.Input?.Mode == Constants.InputMode_Batching)
```
[the remark this answers is not in the document]
Comment (kjcho-msft) [reply Services/DataX.Config/DataX.Config.Input.EventHub/Processor/CreateEventHubConsumerGroup.cs]: fixed. moved to a new place named ResolveBatchInput.

File: DeploymentCloud/Deployment.Common/CosmosDB/commons.json [hunk 2]
```diff
@@ -701,5 +717,171 @@
...
+				"name" : "${sparkJobName}",
+				"file" : "wasbs:///datax/bin/datax-host_2.3_2.11-1.1.0-SNAPSHOT.jar",
+				"className" : "datax.app.BatchApp",
+				"args" : [ 
+					"conf=${sparkJobConfigFilePath}", 
+					"driverLogLevel=${sparkJobDriverLogLevel}"
+				],
+				"jars" : [ 
+					"wasbs:///datax/bin/datax-core_2.3_2.11-1.1.0-SNAPSHOT.jar", 
+					"wasbs:///datax/bin/datax-utility_2.3_2.11-1.1.0-SNAPSHOT.jar",
```
Comment (carlbrochu) [inline DeploymentCloud/Deployment.Common/CosmosDB/commons.json]: We'll need to remove the snapshots when going to stable.  Can you take a note

File: DeploymentCloud/Deployment.Common/CosmosDB/commons.json [hunk 3]
```diff
@@ -701,5 +717,171 @@
...
+				  "spark\uff0edriver\uff0euserClassPathFirst": true,
+				  "spark\uff0eexecutor\uff0euserClassPathFirst": true,
+				  "spark\uff0especulation": true,
+				  "spark\uff0especulation\uff0emultiplier": 3,
+				  "spark\uff0especulation\uff0equantile": 0.9,
+				  "spark\uff0eexecutorEnv\uff0eDATAX_DEFAULTVAULTNAME" : "$sparkKVName",
+				  "spark\uff0eexecutorEnv\uff0eDATAX_APPINSIGHT_KEY": "$appinsightkey",
+				  "spark\uff0eexecutorEnv\uff0eDATAX_APPNAME": "${sparkJobName}",
+				  "spark\uff0eexecutorEnv\uff0eDATAX_LOGLEVEL": "${sparkJobExecutorLogLevel}",
+				  "spark\uff0eexecutorEnv\uff0eDATAX_CHECKPOINTENABLED": "true",
```
Comment (carlbrochu) [inline DeploymentCloud/Deployment.Common/CosmosDB/commons.json]: \ff0e characters seem weird here.  Should this be a \?

File: Services/DataX.Config/DataX.Config.Input.EventHub/Processor/CreateEventHubConsumerGroup.cs [hunk 4]
```diff
@@ -77,15 +77,6 @@ public override async Task<FlowGuiConfig> HandleSensitiveData(FlowGuiConfig guiC
                 guiConfig.Input.Properties.InputResourceGroup = secretId;
             }
 
-            // Replace Info Databricks Token
-            var infoDatabricksToken = guiConfig?.DatabricksToken;
-            if(!string.IsNullOrEmpty(infoDatabricksToken) && !KeyVaultUri.IsSecretUri(infoDatabricksToken))
-            {
-                var secretName = $"{guiConfig.Name}-info-databricksToken";
-                var secretId = await KeyVaultClient.SaveSecretAsync(runtimeKeyVaultName, secretName, infoDatabricksToken, Configuration[Constants.ConfigSettingName_SparkType]);
```
Comment (carlbrochu) [inline Services/DataX.Config/DataX.Config.Input.EventHub/Processor/CreateEventHubConsumerGroup.cs]: where did this go?

File: Services/DataX.Flow/DataX.Flow.Scheduler/TimedScheduler.cs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+               _logger.LogInformation($"{DateTime.UtcNow}: TimedScheduler background task is stopping."));
+
+            while (!stoppingToken.IsCancellationRequested)
+            {
+                _logger.LogInformation($"{DateTime.UtcNow}:TimedScheduler task doing background work.");
+
+                await StartBatchJobs();
+
+                await Task.Delay(20 * 1000, stoppingToken);
+                //await Task.Delay(_schedulerWakeupFrequencyInMin * _oneMinInMilliSeconds, stoppingToken);
```
Comment (carlbrochu) [inline Services/DataX.Flow/DataX.Flow.Scheduler/TimedScheduler.cs]: should this be 20s?

File: Services/DataX.Flow/DataX.Utility.ServiceCommunication/InterServiceCommunicator.cs [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+using System.Threading.Tasks;
+
+namespace DataX.Utility.ServiceCommunication
+{
+    /// <summary>
+    /// Concrete implementation of inter-service communication client.
+    /// </summary>
+    public class InterServiceCommunicator
+    {
+        private const string _ReverseProxyPort = "19081";
```
Comment (carlbrochu) [inline Services/DataX.Flow/DataX.Utility.ServiceCommunication/InterServiceCommunicator.cs]: does this have to be hardcoded?

File: Services/DataX.Flow/DataX.Utility.ServiceCommunication/InterServiceCommunicator.cs [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+                throw;
+            }
+
+            var result = new ApiResult();
+            ProxyResponse(result, response);
+
+            return result;
+        }
+
+        private async void ProxyResponse(ApiResult result, HttpResponseMessage response)
```
Comment (carlbrochu) [inline Services/DataX.Flow/DataX.Utility.ServiceCommunication/InterServiceCommunicator.cs]: headers

File: Services/DataX.Flow/DataX.Flow.Scheduler/TimedScheduler.cs (hunk 5 above)
Comment (carlbrochu) [reply Services/DataX.Flow/DataX.Flow.Scheduler/TimedScheduler.cs]: seems fixed now

File: DeploymentCloud/Deployment.Common/CosmosDB/commons.json (hunk 2 above)
Comment (kjcho-msft) [reply DeploymentCloud/Deployment.Common/CosmosDB/commons.json]: Sure, yes it will be a part of the release process.

Comment (kjcho-msft) [reply DeploymentCloud/Deployment.Common/CosmosDB/commons.json]: now I remembered that even for Master, we shouldn't have the snapshots. will update them.

File: DeploymentCloud/Deployment.Common/CosmosDB/commons.json (hunk 3 above)
Comment (kjcho-msft) [reply DeploymentCloud/Deployment.Common/CosmosDB/commons.json]: It should be encoded this way to convert a bson document for MongoDB,

File: Services/DataX.Config/DataX.Config.Input.EventHub/Processor/CreateEventHubConsumerGroup.cs (hunk 4 above)
Comment (kjcho-msft) [reply Services/DataX.Config/DataX.Config.Input.EventHub/Processor/CreateEventHubConsumerGroup.cs]: it's removed by accident. will get this back.

File: Services/DataX.Flow/DataX.Flow.Scheduler/TimedScheduler.cs (hunk 5 above)
Comment (kjcho-msft) [reply Services/DataX.Flow/DataX.Flow.Scheduler/TimedScheduler.cs]: only test purpose

File: Services/DataX.Flow/DataX.Utility.ServiceCommunication/InterServiceCommunicator.cs (hunk 7 above)
Comment (kjcho-msft) [reply Services/DataX.Flow/DataX.Utility.ServiceCommunication/InterServiceCommunicator.cs]: added

File: Services/DataX.Flow/DataX.Flow.Scheduler/TimedScheduler.cs (hunk 5 above)
Comment (kjcho-msft) [reply Services/DataX.Flow/DataX.Flow.Scheduler/TimedScheduler.cs]: updated.

File: Services/DataX.Flow/DataX.Utility.ServiceCommunication/InterServiceCommunicator.cs (hunk 6 above)
Comment (kjcho-msft) [reply Services/DataX.Flow/DataX.Utility.ServiceCommunication/InterServiceCommunicator.cs]: Yes this is the port that has been specified for the reverse proxy. Gateway uses the same.

File: Services/DataX.Config/DataX.Config/ConfigGeneration/Processor/S650_FlattenJobConfig.cs [hunk 8]
```diff
@@ -54,20 +60,122 @@ public override async Task<string> Process(FlowDeploymentSession flowToDeploy)
...
+
+                    }
                 }
             }
-            
+
             return "done";
         }
+
+        private async Task<string> GetBatchConfigContent(FlowGuiConfig inputConfig, JobConfig jc, JobDeploymentSession job)
```
[the remark this answers is not in the document]
Comment (kjcho-msft) [reply Services/DataX.Config/DataX.Config/ConfigGeneration/Processor/S650_FlattenJobConfig.cs]: This particular function moved to GenerateJobConfigBatch.cs and added headers in the file.

File: Services/DataX.Config/DataX.Config/ConfigGeneration/Processor/S500_ResolveOutputs.cs [hunk 9]
```diff
@@ -186,15 +186,57 @@ private async Task<FlowOutputSpec[]> ProcessOutputs(FlowGuiOutput[] uiOutputs, R
             return fOutputList.ToArray();
         }
 
-        private async Task<FlowBlobOutputSpec> ProcessOutputBlob(string configName, FlowGuiOutput uiOutput)
+        private string TransformPartitionFormat(string partitionFormat)
```
[the remark this answers is not in the document]
Comment (kjcho-msft) [reply Services/DataX.Config/DataX.Config/ConfigGeneration/Processor/S500_ResolveOutputs.cs]: yes it's tested.

File: DeploymentCloud/Deployment.Common/CosmosDB/commons.json (hunk 2 above)
Comment (carlbrochu) [reply DeploymentCloud/Deployment.Common/CosmosDB/commons.json]: @rohit489 can you confirm?

Comment (rohit489) [reply DeploymentCloud/Deployment.Common/CosmosDB/commons.json]: ARM template should not include snapshots as even for master branch we fetch packages from our public nuget feed which

File: Services/DataX.Flow/DataX.Flow.Scheduler/DataX.Flow.Scheduler.csproj [hunk 10]
```diff
@@ -0,0 +1,16 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netcoreapp2.1</TargetFramework>
```
[the remark this answers is not in the document]
Comment (kjcho-msft) [reply Services/DataX.Flow/DataX.Flow.Scheduler/DataX.Flow.Scheduler.csproj]: done

**Pull request merged**